File: src/NodePilot.Core/Activities/KnownProgramLaunchers.cs
namespace NodePilot.Core.Activities;

/// <summary>
/// The handful of Windows launchers a `startProgram` author may name without a directory.
///
/// <para>`filePath` must be a fully qualified path — the engine launches through CreateProcess and
/// never searches the target's PATH, so a bare name would resolve differently per machine and could
/// not be checked against <c>FileSystemOperation:AllowedRoots</c>. These four are completed to their
/// system locations instead of being rejected, both on SCOrch import and while authoring in the
/// designer, so the stored definition always carries the absolute path.</para>
///
/// <para>Mirrored in <c>src/nodepilot-ui/src/lib/knownProgramLaunchers.ts</c>; the two are pinned
/// together by <c>KnownProgramLaunchersFrontendSyncTests</c>.</para>
/// </summary>
public static class KnownProgramLaunchers
{
    public const string CmdExe = @"C:\Windows\System32\cmd.exe";
    public const string PowerShellExe = @"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe";
    public const string CScriptExe = @"C:\Windows\System32\cscript.exe";
    public const string WScriptExe = @"C:\Windows\System32\wscript.exe";

    /// <summary>Launcher stem (no extension) to its absolute path.</summary>
    public static readonly IReadOnlyDictionary<string, string> ByName =
        new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["cmd"] = CmdExe,
            ["powershell"] = PowerShellExe,
            ["cscript"] = CScriptExe,
            ["wscript"] = WScriptExe,
        };

    /// <summary>
    /// Completes a bare launcher name to its absolute path. Returns false for anything that
    /// carries a directory of its own, so a deliberate path to another copy stays untouched.
    /// </summary>
    public static bool TryResolve(string? program, out string resolved)
    {
        resolved = string.Empty;
        if (string.IsNullOrWhiteSpace(program)) return false;

        var value = program.Trim();
        if (value.Contains('\\') || value.Contains('/') || value.Contains(':')) return false;

        var stem = Path.GetFileNameWithoutExtension(value);
        if (!ByName.TryGetValue(stem, out var path)) return false;

        resolved = path;
        return true;
    }
}

File: src/nodepilot-ui/src/lib/knownProgramLaunchers.ts
/**
 * Mirror of `NodePilot.Core.Activities.KnownProgramLaunchers`.
 *
 * `startProgram` requires a fully qualified `filePath`: the engine launches through CreateProcess
 * and never searches the target's PATH, so a bare name would resolve differently per machine and
 * could not be checked against `FileSystemOperation:AllowedRoots`. These four launchers are
 * completed to their system locations instead of being rejected, so an author can still type
 * `cmd.exe` and the stored definition carries the absolute path.
 *
 * Pinned to the backend by `KnownProgramLaunchersFrontendSyncTests`.
 */
export const KNOWN_PROGRAM_LAUNCHERS: Readonly<Record<string, string>> = {
  cmd: 'C:\\Windows\\System32\\cmd.exe',
  powershell: 'C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe',
  cscript: 'C:\\Windows\\System32\\cscript.exe',
  wscript: 'C:\\Windows\\System32\\wscript.exe',
};

/**
 * Completes a bare launcher name to its absolute path. Returns null for anything carrying a
 * directory of its own, so a deliberate path to another copy stays untouched.
 */
export function resolveKnownLauncher(program: string): string | null {
  const value = program.trim();
  if (!value || value.includes('\\') || value.includes('/') || value.includes(':')) return null;
  const stem = value.replace(/\.[^.]*$/, '').toLowerCase();
  return KNOWN_PROGRAM_LAUNCHERS[stem] ?? null;
}

/** True for `C:\dir\file.exe` — the only shape the engine's path guard accepts. */
export function isFullyQualifiedLocalPath(value: string): boolean {
  return /^[A-Za-z]:[\\/]/.test(value.trim());
}

/** True for `\\server\share\...`, which the engine rejects outright (SMB coercion). */
export function isUncPath(value: string): boolean {
  const trimmed = value.trim();
  return trimmed.startsWith('\\\\') || trimmed.startsWith('//');
}
